src/web-extensions.c

```diff
@@ -0,0 +1,696 @@
+      if (json_manifest != NULL)
+          *json_manifest = json_to_string (json_parser_get_root (parser), FALSE);
+      return g_strdup (json_object_get_string_member (metadata_root, "path"));
+    }
+
+  g_set_error (error, XDG_DESKTOP_PORTAL_ERROR, XDG_DESKTOP_PORTAL_ERROR_NOT_FOUND, "cannot find native messaging server");
+  return NULL;
+}
+
+static gboolean
+handle_get_manifest (XdpDbusWebExtensions *object,
+                     GDBusMethodInvocation *invocation,
+                     const char *arg_session_handle,
+                     const char *arg_name,
+                     const char *arg_extension_or_origin)
+{
+  Call *call = call_from_invocation (invocation);
+  Session *session;
+  WebExtensionsSession *we_session;
+  g_autofree char *server_path = NULL;
+  g_autofree char *json_manifest = NULL;
+  g_autoptr(GError) error = NULL;
+
+  session = acquire_session_from_call (arg_session_handle, call);
+  if (!session)
+    {
+      g_dbus_method_invocation_return_error (invocation,
+                                             G_DBUS_ERROR,
+                                             G_DBUS_ERROR_ACCESS_DENIED,
+                                             "Invalid session");
+      return TRUE;
+    }
+
+  SESSION_AUTOLOCK_UNREF (session);
+  we_session = (WebExtensionsSession *)session;
+
+  if (we_session->state != WEB_EXTENSIONS_SESSION_STATE_INIT)
+    {
+      g_dbus_method_invocation_return_error (invocation,
+                                             G_DBUS_ERROR,
+                                             G_DBUS_ERROR_FAILED,
+                                             "Session already started");
+      return TRUE;
+    }
+
+  server_path = find_server (arg_name, arg_extension_or_origin,
+                             NULL, &json_manifest, &error);
+  if (!server_path)
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+
+  xdp_dbus_web_extensions_complete_get_manifest (object, invocation, json_manifest);
+  return TRUE;
+}
+
+static void
+handle_start_in_thread (GTask *task,
+                        gpointer source_object,
+                        gpointer task_data,
+                        GCancellable *cancellable)
+{
+  Request *request = (Request *)task_data;
+  Session *session;
+  WebExtensionsSession *we_session;
+  const char *arg_name;
+  const char *arg_extension_or_origin;
+  const char *app_id;
+  g_autofree char *server_path = NULL;
+  g_autofree char *server_description = NULL;
+  guint response = XDG_DESKTOP_PORTAL_RESPONSE_OTHER;
+  gboolean should_close_session;
+  Permission permission;
+  gboolean allowed;
+  char *argv[] = {NULL, NULL};
+  g_autoptr(GError) error = NULL;
+
+  REQUEST_AUTOLOCK (request);
+  session = g_object_get_data (G_OBJECT (request), "session");
+  SESSION_AUTOLOCK_UNREF (g_object_ref (session));
+  g_object_set_data (G_OBJECT (request), "session", NULL);
+  we_session = (WebExtensionsSession *)session;
```

Comment:
I think I'd prefer `web_extension_session` as the variable name - the `we` is confusing me slightly